OpenFOAM container example (#30)
ubccr/ccr-examples: containers/2_ApplicationSpecific/OpenFOAM/EXAMPLES.md

# OpenFOAM Examples

Please Note: Graphical output (paraFoam) will only display in an [OnDemand portal](https://ondemand.ccr.buffalo.edu) session

## Text only Example

Start an interactive job e.g.

```
salloc --cluster=ub-hpc --partition=general-compute --qos=general-compute --mem=128GB --nodes=1 --cpus-per-task=1 --tasks-per-node=32 --time=01:00:00
```

...then run the container

```
cd /projects/academic/[YourGroupName]/OpenFOAM
apptainer shell -B /util:/util,/scratch:/scratch,/projects:/projects OpenFOAM-13-$(arch).sif
```

sample output:

> ```
> Run the following to use OpenFOAM:
>
>
> source /opt/openfoam13/etc/bashrc && mkdir -p "${FOAM_RUN}" "${FOAM_JOB_DIR}"
>
> Apptainer>
> ```

All the following commands are run from the "Apptainer> " prompt

> Copy the line from above and to use OpenFOAM

Reviewer: ```suggestion
Copy the line from above to use OpenFOAM
```

```
source /opt/openfoam13/etc/bashrc && mkdir -p "${FOAM_RUN}" "${FOAM_JOB_DIR}"
```

Example Run

> Taken from the OpenFOAM 13 [Backward-facing step] https://doc.cfd.direct/openfoam/user-guide-v13/backwardstep) example

Reviewer: ```suggestion
Taken from the OpenFOAM 13 [Backward-facing step](https://doc.cfd.direct/openfoam/user-guide-v13/backwardstep) example
```

> 
> // * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * //
> [...]
> Time = 285s
> 
> smoothSolver:  Solving for Ux, Initial residual = 0.000122524, Final residual = 1.21333e-05, No Iterations 5
> smoothSolver:  Solving for Uy, Initial residual = 0.000992942, Final residual = 6.75593e-05, No Iterations 6
> GAMG:  Solving for p, Initial residual = 0.00123295, Final residual = 0.000112114, No Iterations 4
> time step continuity errors : sum local = 0.0047489, global = 0.000564888, cumulative = 0.55796
> smoothSolver:  Solving for epsilon, Initial residual = 0.000139397, Final residual = 8.97324e-06, No Iterations 3
> smoothSolver:  Solving for k, Initial residual = 0.000221974, Final residual = 1.35975e-05, No Iterations 4
> ExecutionTime = 7.80921 s  ClockTime = 9 s
> 
> 
> SIMPLE solution converged in 285 iterations
> 
> streamlines streamlines write:
>     Seeded 10 particles
>     Sampled 12040 locations
> 
> End
> 
> ```

## ParaFOAM

Since paraFoam is a GUI application, it must be run in an [OnDemand portal](https://ondemand.ccr.buffalo.edu) session
and generally requires at least 128GB of memory to run

> For exampple:

Reviewer: ```suggestion
For example:
```

Open a browser window to our [OnDemand portal](https://ondemand.ccr.buffalo.edu)

[UB-HPC & Faculty Cluster Desktop]

Cluster: UB-HPC  
Slurm Account: [Your Slurm account]  
Partition: [general-compute]  
Quality of Service: [general-compute]  
Number of Hours Requested: 4
Number of Cores: 28
Amount of Memory: 128000

[Launch]

Once the Slurm job starts you can press the button that appears
[Launch UB-HPC & Faculty Cluster Desktop]

A new window will open with the GUI displayed.

Open a terminal with:
[Applications][Terminal Emulator]

cd to your OpenFOAM directory

```
cd /projects/academic/[YourGroupName]/OpenFOAM
```

...and start the container shell
NOTE: This is NOT the same command as above, there are added bind mounts that

>  are neccessary to run paraFoam in CCR's OnDemand

Reviewer: ```suggestion
 are necessary to run paraFoam in CCR's OnDemand
```

```
apptainer shell -B /util:/util -B /scratch:/scratch -B /projects:/projects \
 -B ${XDG_RUNTIME_DIR}:${XDG_RUNTIME_DIR} -B /cvmfs:/cvmfs ./OpenFOAM-13-$(arch).sif
```

sample output:

> ```
> Run the following to use OpenFOAM:
> 
> 
> source /opt/openfoam13/etc/bashrc && mkdir -p "${FOAM_RUN}" "${FOAM_JOB_DIR}"
> 
> Apptainer> 
> ```

All the following commands are run from the "Apptainer> " prompt

copy the line from above to use OpenFOAM

```
source /opt/openfoam13/etc/bashrc && mkdir -p "${FOAM_RUN}" "${FOAM_JOB_DIR}"
```

> This example uses the data gerated in the text example above, following

Reviewer: ```suggestion
This example uses the data generated in the text example above, following
```

> the [Backward-facing step] https://doc.cfd.direct/openfoam/user-guide-v13/backwardstep) example

Reviewer: ```suggestion
the [Backward-facing step](https://doc.cfd.direct/openfoam/user-guide-v13/backwardstep) example
```

Note that those commands could have been run from the "Apptainer> " prompt here.

```
cd "${FOAM_RUN}/pitzDailySteady"
paraFoam
```

This will start with a ParaView splash screen - press the [Close] button.
In the lower left hand window pane scroll down to "Mesh Parts" and select

> the dark buttin next to the left of the words "Mesh Parts" then prss [Apply]
> you will then see the mesh as a sold block of one color

Reviewer: ```suggestion
the dark button next to the left of the words "Mesh Parts" then press [Apply]
you will then see the mesh as a solid block of one color
```
